crates/ironrdp-cliprdr-format/src/html.rs

```diff
@@ -0,0 +1,144 @@
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+pub enum HtmlError {
+    #[error("Invalid CF_HTML format")]
+    InvalidFormat,
+    #[error("Invalid UTF-8")]
+    InvalidUtf8(#[from] std::string::FromUtf8Error),
+    #[error("Failed to parse integer")]
+    InvalidInteger(#[from] std::num::ParseIntError),
+    #[error("Invalid integer conversion")]
+    InvalidConversion,
+}
+
+/// Convert `CF_HTML` format to plain text.
+pub fn cf_html_to_text(input: &[u8]) -> Result<String, HtmlError> {
+    let mut start_fragment = None;
+    let mut end_fragment = None;
+
+    let mut headers_cursor = input;
+
+    let fragment = loop {
+        // Line split logic is manual instead of using BufReader::read_line because
+        // the line ending could be represented as `\r\n`, `\n` or even `\r`.
+        const ENDLINE_CONTROLS: &[u8] = &[b'\r', b'\n'];
+
+        // Failed to find the end of the line
+        let end_pos = match headers_cursor.iter().position(|ch| ENDLINE_CONTROLS.contains(ch)) {
+            Some(pos) => pos,
+            None => return Err(HtmlError::InvalidFormat),
+        };
+
+        let line = String::from_utf8(headers_cursor[..end_pos].to_vec())?;
+
+        let header_value_to_u32 = |value: &str| value.trim_start_matches('0').parse::<u32>();
+
+        match line.split_once(':') {
+            Some((key, value)) => match key {
+                "StartFragment" => {
+                    start_fragment = Some(header_value_to_u32(value)?);
+                }
+                "EndFragment" => {
+                    end_fragment = Some(header_value_to_u32(value)?);
+                }
+                _ => {
+                    // We are not interested in other headers.
+                }
+            },
+            None => {
+                if start_fragment.is_none() || end_fragment.is_none() {
+                    // We reached the end of the headers, but we didn't find the required ones,
+                    // so the format is invalid.
+                    return Err(HtmlError::InvalidFormat);
+                }
+            }
+        };
+
+        if let (Some(start), Some(end)) = (start_fragment, end_fragment) {
+            let start = usize::try_from(start).map_err(|_| HtmlError::InvalidConversion)?;
+            let end = usize::try_from(end).map_err(|_| HtmlError::InvalidConversion)?;
+
+            // Extract fragment from the original buffer.
+            if start > end || end > input.len() {
+                return Err(HtmlError::InvalidFormat);
+            }
+
+            break String::from_utf8(input[start..end].to_vec())?;
+        }
+
+        // INVARIANT: end_pos < headers_cursor.len() - 1
+        #[allow(clippy::arithmetic_side_effects)]
```

Comment:
The `INVARIANT:` comment should come where `end_pos` is declared so the reader knows what is being enforced for this value before reading the logic itself.

```rust
// INVARIANT: end_pos < headers_cursor.len() - 1
let end_pos = /* … */;
```

Example in rust-analyzer code:
https://github.com/rust-lang/rust-analyzer/blob/c1c9e10f72ffd2e829d20ff1439ff49c2e121731/crates/hir-ty/src/lower.rs#L1118-L1127

When it’s not obvious why some piece of code is indeed ensuring the invariant holds, a comment may be added:
https://github.com/rust-lang/rust-analyzer/blob/c1c9e10f72ffd2e829d20ff1439ff49c2e121731/crates/hir-ty/src/lower.rs#L1143-L1144

Then, the invariant can be used to explain why it’s okay to do something non obvious such as disabling a lint:
```rust
#[allow(clippy::arithmetic_side_effects)] // given the invariant on `end_pos`, this code can’t cause integer overflow
{
    /* … */
}
```